include/net/wifi_mgmt.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2017 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief WiFi L2 stack public header
+ */
+
+#ifndef __WIFI_MGMT_H__
+#define __WIFI_MGMT_H__
+
+#include <net/net_mgmt.h>
+
+/* Management part definitions */
+
+#define _NET_WIFI_LAYER	NET_MGMT_LAYER_L2
+#define _NET_WIFI_CODE	0x156
+#define _NET_WIFI_BASE	(NET_MGMT_IFACE_BIT |			\
+			 NET_MGMT_LAYER(_NET_WIFI_LAYER) |	\
+			 NET_MGMT_LAYER_CODE(_NET_WIFI_CODE))
+#define _NET_WIFI_EVENT	(_NET_WIFI_BASE | NET_MGMT_EVENT_BIT)
+
+enum net_request_wifi_cmd {
+	NET_REQUEST_WIFI_CMD_SCAN = 1,
+	NET_REQUEST_WIFI_CMD_CONNECT,
+	NET_REQUEST_WIFI_CMD_DISCONNECT,
+};
+
+#define NET_REQUEST_WIFI_SCAN					\
+	(_NET_WIFI_BASE | NET_REQUEST_WIFI_CMD_SCAN)
+
+NET_MGMT_DEFINE_REQUEST_HANDLER(NET_REQUEST_WIFI_SCAN);
+
+#define NET_REQUEST_WIFI_CONNECT				\
+	(_NET_WIFI_BASE | NET_REQUEST_WIFI_CMD_CONNECT)
+
+NET_MGMT_DEFINE_REQUEST_HANDLER(NET_REQUEST_WIFI_CONNECT);
+
+#define NET_REQUEST_WIFI_DISCONNECT				\
+	(_NET_WIFI_BASE | NET_REQUEST_WIFI_CMD_DISCONNECT)
+
+NET_MGMT_DEFINE_REQUEST_HANDLER(NET_REQUEST_WIFI_DISCONNECT);
+
+enum net_event_wifi_cmd {
+	NET_EVENT_WIFI_CMD_SCAN_RESULT = 1,
+	NET_EVENT_WIFI_CMD_CONNECT_RESULT,
+	NET_EVENT_WIFI_CMD_DISCONNECT_RESULT,
+};
+
+#define NET_EVENT_WIFI_SCAN_RESULT				\
+	(_NET_WIFI_EVENT | NET_EVENT_WIFI_CMD_SCAN_RESULT)
+
+#define NET_EVENT_WIFI_CONNECT_RESULT				\
+	(_NET_WIFI_EVENT | NET_EVENT_WIFI_CMD_CONNECT_RESULT)
+
+#define NET_EVENT_WIFI_DISCONNECT_RESULT			\
+	(_NET_WIFI_EVENT | NET_EVENT_WIFI_CMD_DISCONNECT_RESULT)
+
+
+#define WIFI_CHANNEL_MAX 14
+#define WIFI_CHANNEL_ANY (SHORT_MAX)
+
+enum wifi_security_type {
+	WIFI_SECURITY_TYPE_NONE = 0
+	WIFI_SECURITY_TYPE_PSK,
+};
+
+/* Each result is provided to the net_mgmt_event_callback
+ * via its info attribute (see net_mgmt.h)
+ */
+struct wifi_scan_result {
+	u8_t ssid[WIFI_SSID_MAX_LEN];
+	u8_t ssid_length;
+
+	u8_t channel;
+	enum wifi_security_type security;
+	u8_t rssi;
+};
+
+struct wifi_connect_req_params {
+	u8_t *ssid;
+	u8_t ssid_length; /* Max 32 */
+
+	u8_t *psk;
+	u8_t psk_length; /* Min 8 - Max 64 */
+
+	u8_t channel;
+	enum wifi_security_type security;
+};
+
+struct wifi_status {
+	int status;
+};
+
+#ifdef CONFIG_WIFI_OFFLOAD
+
+#include <net/net_if.h>
+
+typedef void (*scan_result_cb_t)(struct net_if *iface,
```

Comment:
We could. But what's the policy of sorting the networks in ext/hal side?  Is it showing only first 10 results it found, is it showing 10 results which have best signal quality, is it callable many times so you can get 10 by 10 results etc... ? I don't see this being easily ported to all ext/hal etc...

Reply:
You can also check Inventek library in my PR #5075. Interface is in wifi.h file.
For scan, user give a table and a max number : 
    WIFI_Status_t       WIFI_ListAccessPoints(WIFI_APs_t *APs, uint8_t AP_MaxNbr);
Inventek documentation is not really detailed, but in my experience, it returns the most powerfull networks first.

Reply:
network as grouped AP serving same SSID/Security, or as for TI: one entry per-AP, even multiple ones serve the same SSID/Security?

Reply:
For TI ext/hal, first the user sets the scan policy (RSSI threshold, automatic scans and seconds between scans, one-shot scans, etc). Max of 30 results are scanned, which from the examples, appears to be sorted by strength.
Then user can ask for number (<= 30) of results.   
User can call WlanGetNetworkList() repeatedly, to trigger new scans (if one-shot policy was set) and get updated list.
I anticipate that we can have a Kconfig option to set some standard policy (like one-shot), and not expose all those configuration APIs to the user to change at run-time, to try to keep things simple.

Reply:
> Otherwise seems the user must wait for multiple callbacks to build his own scan results table, and will not know when the last callback arrives.

The problem with the pre-allocated list is that it needs memory to be allocated either from stack or statically. We could add a parameter to scan_cb that tells if this is the last result or not.
Would it be possible to have a scan_cb that could have things in both ways, so if the user provides an array, then that is used to populate the result, and if the array is omitted, then callback would be called for each result?

Reply:
> Would it be possible to have a scan_cb that could have things in both ways
That would be nice, as the offload device code could be simplified, and not be forced to loop through its internal table of scan results calling the scan_cb multiple times.